src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/Renamer.cs

```diff
@@ -78,18 +83,131 @@ public async void OnWorkspaceChangedAsync(object sender, WorkspaceChangeEventArg
+            }
+        }
+
+        private static async Task<bool> ShouldRenameAsync(Project project, string oldNameBase, string newFilePath, bool isCaseSensitive)
+        {
+            // Do not rename is the name does not change
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+            if (string.Compare(oldNameBase, newName, !isCaseSensitive) == 0)
+                return false;
+
+            // Do not rename if we cannot find the symbol
+            ISymbol symbol = await TryGetSymbolToRenameAsync(project, oldNameBase, newFilePath, isCaseSensitive);
+            return symbol != null;
+        }
+
+        private static async Task<ISymbol> TryGetSymbolToRenameAsync(Project myNewProject, string oldNameBase, string newFilePath, bool isCaseSensitive)
+        {
+            Project project = myNewProject;
+
+            Document newDocument = GetDocument(project, newFilePath);
+            if (newDocument == null)
+                return null;
+
+            SyntaxNode root = await GetRootNode(newDocument);
+            if (root == null)
+                return null;
+
+            SemanticModel semanticModel = await newDocument.GetSemanticModelAsync();
+            if (semanticModel == null)
+                return null;
+
+            IEnumerable<SyntaxNode> declarations = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(semanticModel, n, oldNameBase, isCaseSensitive));
+            SyntaxNode declaration = declarations.FirstOrDefault();
+            if (declaration == null)
+                return null;
+
+            return semanticModel.GetDeclaredSymbol(declaration);
+        }
+
+        private async Task<Solution> GetRenamedSolutionAsync(Project myNewProject, string oldNameBase, string newFilePath, bool isCaseSensitive)
+        {
+            Project project = myNewProject;
+            ISymbol symbol = await TryGetSymbolToRenameAsync(myNewProject, oldNameBase, newFilePath, isCaseSensitive);
+            if (symbol == null)
+                return null;
+
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+            Solution solutionToRename = null;
+            while (project != null)
+            {
+                // Note that RenameSymbolAsync will return a new snapshot of solution.
+                solutionToRename = await _roslynServices.RenameSymbolAsync(project.Solution, symbol, newName);
+                project = solutionToRename.Projects.Where(p => StringComparers.Paths.Equals(p.FilePath, myNewProject.FilePath)).FirstOrDefault();
+            }
+
+            return solutionToRename;
+        }
+
+        private static Document GetDocument(Project project, string filePath) =>
+            (from d in project.Documents where StringComparers.Paths.Equals(d.FilePath, filePath) select d).FirstOrDefault();
+
+        private static Task<SyntaxNode> GetRootNode(Document newDocument) =>
+            newDocument.GetSyntaxRootAsync();
```

Comment:
This method seems unnecessary.

Reply:
This was in the previous implementation.  I will remove/simplify this logic in a future PR